### PR TITLE
Fix type annotation properties assigning to undefined. (#8417)

### LINKED ISSUE #8417: v7 plugin-proposal-class-properties transform flow type annotation to undefined
## Bug Report

**Current Behavior**

`@babel/plugin-proposal-class-properties` transform flow type annotation to a property assigning undefined.

this may caught statement order unexpected error.

**Input Code**

```js
let code = `
class B {}
class Logger extends B {
  log: {
    error: boolean
  }
  silent: boolean
  log = {
    error: true
  }
  get silent(): boolean {
    return this._silent;
  }
  set silent(val: boolean) {
    this._silent = !!val;
    if (!this.log) {
      throw new Error('#log should not be nil.')
    }
    Object.keys(this.log).forEach(k => { /* do something... */  })
  }
  constructor() {
    super()
  }
}
`

let r = require("@babel/core").transform(code, {
  "presets": [ "@babel/preset-flow" ],
  "plugins": [
    [ "@babel/plugin-proposal-class-properties", { "loose": true } ]
  ]
})
console.log(r.code)
```

ouput =>

```js
class B {}

class Logger extends B {
  get silent() {
    return this._silent;
  }

  set silent(val) {
    this._silent = !!val;

    if (!this.log) {
      throw new Error('#log should not be nil.')   // <----- will throw if the annotation property assigning undefined
    }

    Object.keys(this.log).forEach(k => { /* do something... */ });
    });
  }

  constructor() {
    super();
    this.log = void 0;  <<<<<----
    this.silent = void 0; <<<<<---- with setter defined.
    this.log = {
      error: true
    };
  }
}
```

**Expected behavior/code**

```js
class B {}

class Logger extends B {
  get silent() {
    return this._silent;
  }

  set silent(val) {
    this._silent = !!val;

    if (!this.log) {
      throw new Error('#log should not be nil.')
    }

    Object.keys(this.log).forEach(k => { /* do something... */ });
  }

  constructor() {
    super();
    this.log = {
      error: true
    };
  }

}
```

**Environment**
- Babel version(s): [e.g. v6.0.0, v7.0.0-beta.55]
- Node/npm version: [e.g. Node 8/npm 5]
- OS: [e.g. OSX 10.13.4]
- Monorepo [e.g. yes/no/Lerna]
- How you are using Babel: `js`

**Possible Solution**

Should strip the `ClassProperty` type of `TypeAnnotation`.

### PR DESCRIPTION
**Pull request by allex**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #8417  
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  | Add a helper `declareNilProps (obj, props);` to @babel/babel-helpers
| License                  | MIT

### CODE REVIEW

File: packages/babel-helpers/src/helpers.js [hunk 1]
```diff
@@ -1053,3 +1053,12 @@ helpers.classPrivateFieldSet = helper("7.0.0-beta.0")`
     return value;
   }
 `;
+
+helpers.declareNilProps = helper("7.0.0-beta.0")`
```
Comment (nicolo-ribaudo) [inline packages/babel-helpers/src/helpers.js]: This should be `7.0.1`, since it represents the first version where the helper is available.

**Comment by nicolo-ribaudo** [comment]
I don't undertand why the `_declareNilProps` helper is needed. Why should `x: number` create a property on the class if it doesn't have an initializer? Isn't it just a type annotation?

**Comment by loganfsmyth** [comment]
I don't see us landing this without a serious discussion with the folks developing the class fields specification and Flow folks. What you're proposing means that if you have

```
class Foo {
  x;
  y;
}

console.log("x" in new Foo());
```
in an existing codebase, and you want to convert the codebase to Flow by doing

```
class Foo {
  x: number;
  y: number;
}

console.log("x" in new Foo());
```
the behavior of the code would change based on the addition of a type annotation, which seems like it would be incredibly surprising.

**Comment by c0d3m3nt0r** [comment]
@loganfsmyth, that scenario you mention makes sense, I mean, if you declared two class variables, and then you do not initialize them, they will remain `undefined`, which is expected, I think, at least I wouldn't expect another behavior in that case.

The thing gets inconsistent, at least from my point of view, when after declaring a variable, like in your example, but having in mind that variable is an arrow function, when it later gets initialized in the code, at the time the constructor gets executed, that variable remains `undefined`, like in [react-native](https://github.com/facebook/react-native/blob/master/Libraries/Animated/src/createAnimatedComponent.js) `createAnimatedComponent()` source code, and as I explained [here](https://github.com/babel/babel/issues/8417#issuecomment-415436804), there is in where things turn out to be inconsistent. 

In my opinion, I believe this is related to the part in where class properties are being processed, I would have expected `@babel/proposal-class-properties` took care of these kind of use cases and output the correct class properties initialization, in where (in the case I pointed) the variable should be other than `undefined`.

The easiest path would be, in my case, to make **react-native** team to consider removing the Flow type annotation for the arrow function at the top of their class, and just initialize the arrow function as they do later in their code (and and the same time make it more accurate, in terms of parameter types and return type), and that should fix the issue, but, the problem is all other libraries that are doing the same thing, having in mind this issue is not happening with **Babel 6.x**.

So, having that in mind, how can we solve this situation?, should we ask everyone to adjust their code in order to workaround this issue? (besides we agree or not that one is a good practice or makes sense or not) or to find a backward compatible solution to make current libraries having the same issue to have their pipeline back to work?

Sooner or later someone will fix this, but in the meantime, the code starts to get full or workarounds to get rid of these kind of issues when moving to **Babel 7**. I have updated to the latest stable version, and I have to recognize several issues were fixed since the `beta.47`, which is a great job to address those issues to finally arrive to something more stable, but this one it is still there, and it would be great to have it fixed at some point :)

Thank you for your support and effort to keep making Babel such a great tool.

File: packages/babel-plugin-proposal-class-properties/test/fixtures/public/type-annotation-transform/exec.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    noInitial: boolean
+
+    log: { error: boolean } = {
+      error: x
+    };
+
+    ver: number
+    ver = x;
+
+    fn: Function
```
[the remark this answers is not in the document]
Comment (c0d3m3nt0r) [reply packages/babel-plugin-proposal-class-properties/test/fixtures/public/type-annotation-transform/exec.js]: Cool, thanks 👍

**Comment by torkelo** [comment]
Any update to get this merged? This is causing issue wtih typescript + babel

**Comment by nicolo-ribaudo** [comment]
I'm closing this PR because you can already use `declare` for class fields with TypeScript to avoid initializing them to `undefined`, and starting from Babel 7.9.0 it will also be supported by Flow (https://github.com/babel/babel/pull/11178).

Thanks @allex anyway!

**Pull request closed without merging**